In [ ]:
import numpy as np
from doctr.io import DocumentFile
from doctr.models import ocr_predictor
from pdf2image import convert_from_path
from PIL import Image, ImageDraw
import matplotlib.pyplot as plt
import json, re

# Step 1: Convert PDF to PIL images
pdf_path = "C:\\Users\\Govind\\Downloads\\Work_And\\Self_Learning\\Invoice_Extraction_demo\\Invoices\\CN-006991.pdf"
# pdf_path1 = "C:\\Users\\Govind\\Downloads\\Work_And\\Self_Learning\\Invoice_Extraction_demo\\Invoices\\B669_CN-016582_20240926_V1127.pdf"
# pdf_path2 = "C:\\Users\\Govind\\Downloads\\Work_And\\Self_Learning\\Invoice_Extraction_demo\\Invoices\\B035_CN-016337_20240906_V1127.pdf"
pil_images = convert_from_path(pdf_path, poppler_path = "C:\\Users\\Govind\\Downloads\\Release-24.08.0-0\\poppler-24.08.0\\Library\\bin")

# Step 2: Convert PIL images to RGB NumPy arrays
image_arrays = [np.array(img.convert("RGB")) for img in pil_images]

# Step 3: Load the OCR model
model = ocr_predictor(pretrained=True)

# Step 4: Run OCR on the image arrays
doc = model(image_arrays)


c:\Users\Govind\Downloads\Work_And\Self_Learning\Invoice_Extraction_demo\CompleteWL\lib\site-packages\tqdm\auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
  from .autonotebook import tqdm as notebook_tqdm
c:\Users\Govind\Downloads\Work_And\Self_Learning\Invoice_Extraction_demo\CompleteWL\lib\site-packages\doctr\models\utils\pytorch.py:59: FutureWarning: You are using `torch.load` with `weights_only=False` (the current default value), which uses the default pickle module implicitly. It is possible to construct malicious pickle data which will execute arbitrary code during unpickling (See https://github.com/pytorch/pytorch/blob/main/SECURITY.md#untrusted-models for more details). In a future release, the default value for `weights_only` will be flipped to `True`. This limits the functions that could be executed during unpickling. Arbitrary objects will no longer be allowed to be loaded v

In [1]:
import numpy as np
from doctr.io import DocumentFile
from doctr.models import ocr_predictor
from pdf2image import convert_from_path
from PIL import Image, ImageDraw
import matplotlib.pyplot as plt
import json, re, os

# Step 1: Convert PDF to PIL images
#pdf_path = "output_text_embedded.pdf"
pdf_path = "C:\\Users\\Govind\\Downloads\\Work_And\\Self_Learning\\Invoice_Extraction_demo\\Invoices\\B669_CN-016582_20240926_V1127.pdf"
pil_images = convert_from_path(pdf_path, poppler_path = "C:\\Users\\Govind\\Downloads\\Release-24.08.0-0\\poppler-24.08.0\\Library\\bin")
output_dir = os.path.join(os.path.dirname(pdf_path), "OCRed")

# Create OCRed folder if it doesn't exist
os.makedirs(output_dir, exist_ok=True)

# Step 2: Convert PIL images to RGB NumPy arrays
image_arrays = [np.array(img.convert("RGB")) for img in pil_images]

# Step 3: Load the OCR model
model = ocr_predictor(pretrained=True)

# Step 4: Run OCR on the image arrays
doc = model(image_arrays)

# Step 5: Visualize + Save OCRed image
for i, (page, pil_img) in enumerate(zip(doc.pages, pil_images)):
    draw = ImageDraw.Draw(pil_img)
    
    for block in page.blocks:
        for line in block.lines:
            # Draw bounding box
            points = (np.array(line.geometry) * [pil_img.width, pil_img.height]).flatten()
            draw.rectangle([points[0], points[1], points[2], points[3]], outline="green", width=2)
            # Optional: draw text near box
            text = " ".join([word.value for word in line.words])
            draw.text((points[0], points[1]-10), text, fill="red")
    
    file_name = pdf_path.split(sep="\\")[-1].rstrip(".pdf")
    output_path = os.path.join(output_dir, f"{file_name}_page_{i+1}_ocred.png")
    pil_img.save(output_path)
    print(f"Saved OCRed image: {output_path}")

# <==========================Required_Fields==============================>
full_text = ""
for page in doc.pages:
    for block in page.blocks:
        for line in block.lines:
            full_text += " ".join([word.value for word in line.words]) + "\n"

c:\Users\Govind\Downloads\Work_And\Self_Learning\Invoice_Extraction_demo\CompleteWL\lib\site-packages\tqdm\auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
  from .autonotebook import tqdm as notebook_tqdm
c:\Users\Govind\Downloads\Work_And\Self_Learning\Invoice_Extraction_demo\CompleteWL\lib\site-packages\doctr\models\utils\pytorch.py:59: FutureWarning: You are using `torch.load` with `weights_only=False` (the current default value), which uses the default pickle module implicitly. It is possible to construct malicious pickle data which will execute arbitrary code during unpickling (See https://github.com/pytorch/pytorch/blob/main/SECURITY.md#untrusted-models for more details). In a future release, the default value for `weights_only` will be flipped to `True`. This limits the functions that could be executed during unpickling. Arbitrary objects will no longer be allowed to be loaded v

Saved OCRed image: C:\Users\Govind\Downloads\Work_And\Self_Learning\Invoice_Extraction_demo\Invoices\OCRed\B669_CN-016582_20240926_V1127_page_1_ocred.png


In [3]:
print(full_text)

INTEGRIA HEALTHCARE (AUSTRALIA) PTY LIMITED
ABN 70096496212
Building 5 Freeway Office Park 2728 Logan Road
Eight Mile Plains QLD 4113
AUS
Ph. 1300 654 336 Fax 1300 654 844
Customer Care Ph. : 1300 654 336
Customer Care Fax : 1300
Email Orders : orders@integria.com
Integria
HEALTHCARE
Credit adjustment note copy
Document No:
Date:
Account No:
Customer Ref:
Delivery No:
Our Order No:
Taken By:
CN-016582
26/09/2024
55327
TT OOD
SO-0318815
Suzanne Aschenbrenner
Ship To: CWH Mount Cotton B669
Payer: CW Management Pty Ltd B655
Building F, 44 Raglan Street
PRESTON VIC 3072
AUS
Sh17,Mount Cotton Central,101 Valle
MOUNT COTTON QLD 4165
AUS
Material
TTAOIL25
Brand
Description
UoM
Ordered Shipped
EA
-1
-1
RRP List Discount Amount Discount % Unit Price Ext Nett
Thursday Plantation Arnica Oil 25ml
29.99 16.85
0.00 0.00% 16.85 -16.85
Accounts Receivable:
Phone:
Fax:
Email:
Notes:
Pay to:
BSB 032-388
Account 133956
Please quote Ref. 55327
30/11/2024
TOTAL ITEMS
Freight
GST
Final Amount
AUD
AUD
-16.85

In [19]:
full_text.replace("\n", " ")

'INTEGRIA HEALTHCARE (AUSTRALIA) PTY LIMITED ABN 70096496212 Building 5 Freeway Office Park 2728 Logan Road Eight Mile Plains QLD 4113 AUS Ph. 1300 654 336 Fax 1300 654 844 Customer Care Ph. : 1300 654 336 Customer Care Fax : 1300 Email Orders : orders@integria.com Integria HEALTHCARE Credit adjustment note copy Document No: Date: Account No: Customer Ref: Delivery No: Our Order No: Taken By: CN-016582 26/09/2024 55327 TT OOD SO-0318815 Suzanne Aschenbrenner Ship To: CWH Mount Cotton B669 Payer: CW Management Pty Ltd B655 Building F, 44 Raglan Street PRESTON VIC 3072 AUS Sh17,Mount Cotton Central,101 Valle MOUNT COTTON QLD 4165 AUS Material TTAOIL25 Brand Description UoM Ordered Shipped EA -1 -1 RRP List Discount Amount Discount % Unit Price Ext Nett Thursday Plantation Arnica Oil 25ml 29.99 16.85 0.00 0.00% 16.85 -16.85 Accounts Receivable: Phone: Fax: Email: Notes: Pay to: BSB 032-388 Account 133956 Please quote Ref. 55327 30/11/2024 TOTAL ITEMS Freight GST Final Amount AUD AUD -16.8

'INTEGRIA HEALTHCARE (AUSTRALIA) PTY LIMITED ABN 70096496212 Building 5 Freeway Office Park 2728 Logan Road Eight Mile Plains QLD 4113 AUS Ph. 1300 654 336 Fax 1300 654 844 Customer Care Ph. : 1300 654 336 Customer Care Fax : 1300 Email Orders : orders@integria.com Integria HEALTHCARE Credit adjustment note copy Document No: Date: Account No: Customer Ref: Delivery No: Our Order No: Taken By: CN-016582 26/09/2024 55327 TT OOD SO-0318815 Suzanne Aschenbrenner Ship To: CWH Mount Cotton B669 Payer: CW Management Pty Ltd B655 Building F, 44 Raglan Street PRESTON VIC 3072 AUS Sh17,Mount Cotton Central,101 Valle MOUNT COTTON QLD 4165 AUS 

Material TTAOIL25 Brand Description UoM Ordered Shipped EA -1 -1 RRP List Discount Amount Discount % Unit Price Ext Nett Thursday Plantation Arnica Oil 25ml 29.99 16.85 0.00 0.00% 16.85 -16.85 

Accounts Receivable: Phone: Fax: Email: Notes: Pay to: BSB 032-388 Account 133956 Please quote Ref. 55327 30/11/2024 TOTAL ITEMS Freight GST Final Amount AUD AUD -16.85 AUD AUD -16.85 0.00 -1.69 -18.54 1300 654 336 1300 06 ar@integria.com 10.00 % Payment Terms: 2 Months from Month End Due date: 1/1 '

In [4]:
full_text

'INTEGRIA HEALTHCARE (AUSTRALIA) PTY LIMITED\nABN 70096496212\nBuilding 5 Freeway Office Park 2728 Logan Road\nEight Mile Plains QLD 4113\nAUS\nPh. 1300 654 336 Fax 1300 654 844\nCustomer Care Ph. : 1300 654 336\nCustomer Care Fax : 1300\nEmail Orders : orders@integria.com\nIntegria\nHEALTHCARE\nCredit adjustment note copy\nDocument No:\nDate:\nAccount No:\nCustomer Ref:\nDelivery No:\nOur Order No:\nTaken By:\nCN-016582\n26/09/2024\n55327\nTT OOD\nSO-0318815\nSuzanne Aschenbrenner\nShip To: CWH Mount Cotton B669\nPayer: CW Management Pty Ltd B655\nBuilding F, 44 Raglan Street\nPRESTON VIC 3072\nAUS\nSh17,Mount Cotton Central,101 Valle\nMOUNT COTTON QLD 4165\nAUS\nMaterial\nTTAOIL25\nBrand\nDescription\nUoM\nOrdered Shipped\nEA\n-1\n-1\nRRP List Discount Amount Discount % Unit Price Ext Nett\nThursday Plantation Arnica Oil 25ml\n29.99 16.85\n0.00 0.00% 16.85 -16.85\nAccounts Receivable:\nPhone:\nFax:\nEmail:\nNotes:\nPay to:\nBSB 032-388\nAccount 133956\nPlease quote Ref. 55327\n30/11/

In [5]:
def extract_invoice_fields(text: str):
    data = {
        "vendor_name": None,
        "store_name": None,
        "store_id": None,
        "bill_to_store_name": None,
        "bill_to_store_id": None,
        "invoice_number": None,
        "date": None,
        "Total_Items": None,
        "Freight": None,
        "Final_Amount": None,
        "GST": None,
        "type": None
    }

    lines = [line.strip() for line in text.splitlines() if line.strip()]
    full_text_clean = " ".join(lines)

    # Vendor
    for line in lines:
        if not data["vendor_name"] and "PTY" in line.upper():
            data["vendor_name"] = line

    # Type
    if "Credit adjustment note" in full_text_clean:
        data["type"] = "Credit adjustment note"

    # Invoice number
    inv_match = re.search(r"(CN-\d+)", full_text_clean)
    if inv_match:
        data["invoice_number"] = inv_match.group(1)

    # Date
    date_match = re.search(r"(\d{2}/\d{2}/\d{4})", full_text_clean)
    if date_match:
        data["date"] = date_match.group(1)

    # Ship To
    ship_match = re.search(r"Ship To:\s*([^\n]+)", text)
    if ship_match:
        ship_line = ship_match.group(1).strip()
        parts = ship_line.split()
        if len(parts) >= 2:
            data["store_id"] = parts[-1]
            data["store_name"] = " ".join(parts[:-1])

    # Payer
    payer_match = re.search(r"Payer:\s*([^\n]+)", text)
    if payer_match:
        payer_line = payer_match.group(1).strip()
        parts = payer_line.split()
        if len(parts) >= 2:
            data["bill_to_store_id"] = parts[-1]
            data["bill_to_store_name"] = " ".join(parts[:-1])

    # Sub total
    sub_match = re.search(r"TOTAL ITEMS.*?AUD\s*([-+]?\d+\.\d{2})", full_text_clean)
    if sub_match:
        data["Total_Items"] = abs(float(sub_match.group(1)))

    # GST
    gst_match = re.search(r"GST.*?(\d+\.\d{2})\s*%", full_text_clean, re.DOTALL)
    if gst_match:
        data["GST"] = abs(float(gst_match.group(1)))

    # Freight
    freight_match = re.search(r"Freight.*?AUD\s*([-+]?\d+\.\d{2})", full_text_clean)
    if freight_match:
        data["Freight"] = abs(float(freight_match.group(1)))

    # Final total
    final_match = re.search(r"Final Amount.*?AUD\s*([-+]?\d+\.\d{2})", full_text_clean)
    if final_match:
        data["Final_Amount"] = abs(float(final_match.group(1)))

    return data
required_fields = extract_invoice_fields(full_text)
import pprint
pprint.pprint(required_fields)


{'Final_Amount': 16.85,
 'Freight': 16.85,
 'GST': 10.0,
 'Total_Items': 16.85,
 'bill_to_store_id': 'B655',
 'bill_to_store_name': 'CW Management Pty Ltd',
 'date': '26/09/2024',
 'invoice_number': 'CN-016582',
 'store_id': 'B669',
 'store_name': 'CWH Mount Cotton',
 'type': 'Credit adjustment note',
 'vendor_name': 'INTEGRIA HEALTHCARE (AUSTRALIA) PTY LIMITED'}


In [6]:
required_fields

{'vendor_name': 'INTEGRIA HEALTHCARE (AUSTRALIA) PTY LIMITED',
 'store_name': 'CWH Mount Cotton',
 'store_id': 'B669',
 'bill_to_store_name': 'CW Management Pty Ltd',
 'bill_to_store_id': 'B655',
 'invoice_number': 'CN-016582',
 'date': '26/09/2024',
 'Total_Items': 16.85,
 'Freight': 16.85,
 'Final_Amount': 16.85,
 'GST': 10.0,
 'type': 'Credit adjustment note'}

In [7]:
re.search(r"Date:.*?(\d{2}/\d{2}/\d{4})", full_text, re.DOTALL).group(1)

'26/09/2024'

In [8]:
pdf_path.split(sep="\\")[-1].rstrip(".pdf")

'B669_CN-016582_20240926_V1127'

In [11]:
def validate_numeric_fields(data):
    issues = []
    
    # Gather numeric fields (ignore None)
    numeric_fields = {k: v for k, v in data.items() if isinstance(v, (int, float)) and v is not None}
    
    # Compare each numeric field with others
    keys = list(numeric_fields.keys())
    for i in range(len(keys)):
        for j in range(i + 1, len(keys)):
            k1, k2 = keys[i], keys[j]
            if numeric_fields[k1] == numeric_fields[k2]:
                issues.append(f"{k1} matches {k2}, suspicious")
                data[k1] = None
                data[k2] = None

    if issues:
        print("⚠️ Post-extraction validation detected issues:")
        for issue in issues:
            print(f" - {issue}")
        print("Fields reset to None. Re-validation required.")
        return True
    else:
        print("✅ Post-extraction validation passed.")
        return False
    
required_fields = extract_invoice_fields(full_text)
needs_validation = validate_numeric_fields(required_fields)

import pprint
pprint.pprint(required_fields)

if needs_validation:
    print("⚠️ Triggering manual validation process...")
    # subprocess.run(["python", "Validation/validation.py"])
else:
    print("✅ Ready for export or next step.")

⚠️ Post-extraction validation detected issues:
 - Total_Items matches Freight, suspicious
 - Total_Items matches Final_Amount, suspicious
 - Freight matches Final_Amount, suspicious
Fields reset to None. Re-validation required.
{'Final_Amount': None,
 'Freight': None,
 'GST': 10.0,
 'Total_Items': None,
 'bill_to_store_id': 'B655',
 'bill_to_store_name': 'CW Management Pty Ltd',
 'date': '26/09/2024',
 'invoice_number': 'CN-016582',
 'store_id': 'B669',
 'store_name': 'CWH Mount Cotton',
 'type': 'Credit adjustment note',
 'vendor_name': 'INTEGRIA HEALTHCARE (AUSTRALIA) PTY LIMITED'}
⚠️ Triggering manual validation process...


In [10]:
# Identify missing fields
missing_fields = [k for k, v in required_fields.items() if not v or str(v).strip() == '']

if missing_fields:
    print("⚠️ Validation required for the following fields:")
    for field in missing_fields:
        print(f" - {field}")

⚠️ Validation required for the following fields:
 - Total_Items
 - Freight
 - Final_Amount


In [ ]:
def load_field_map(json_path="FieldLearningMap.json"):
    if os.path.exists(json_path):
        return json.load(open(json_path))
    return {}

def extract_text_by_coords(image, coords):
    x, y, w, h = coords["x"], coords["y"], coords["width"], coords["height"]
    cropped = image.crop((x, y, x + w, y + h))
    return model([np.array(cropped.convert("RGB"))]).pages[0].export()['blocks'][0]['lines']

def Extract_using_coordinates(required_fields, pdf_path):
    vendor = required_fields.get("vendor_name")
    if not vendor:
        print("❌ Vendor name missing. Cannot proceed.")
        return required_fields

    field_map = load_field_map()
    if vendor not in field_map:
        print(f"❌ No template found for vendor: {vendor}")
        return required_fields

    # Convert PDF to image (first page only)
    images = convert_from_path(pdf_path, poppler_path=r"C:\Users\Govind\Downloads\Release-24.08.0-0\poppler-24.08.0\Library\bin")
    image = images[0]

    field_regex = {
        "invoice_number": r"(CN-\d+)",
        "date": r"(\d{2}/\d{2}/\d{4})",
        "Total_Items": r"([-+]?\d+\.\d{2})",
        "Freight": r"([-+]?\d+\.\d{2})",
        "Final_Amount": r"([-+]?\d+\.\d{2})",
        "GST": r"(\d+\.\d{2})\s*%",
    }

    for field, value in required_fields.items():
        if value is not None or field == "vendor_name":
            continue  # Already extracted or not applicable

        found = False
        print(f"🔍 Trying to auto-extract missing field: {field}")
        for template_id, fields in field_map[vendor].items():
            if field not in fields:
                continue

            coords = fields[field]
            try:
                ocr_lines = extract_text_by_coords(image, coords)
                combined_text = " ".join(word['value'] for line in ocr_lines for word in line['words'])

                pattern = field_regex.get(field)
                if pattern:
                    match = re.search(pattern, combined_text)
                    if match:
                        val = match.group(1)
                        required_fields[field] = val if field in ["invoice_number", "date"] else abs(float(val))
                        print(f"✅ Extracted {field}: {required_fields[field]} using template: {template_id}")
                        found = True
                        break
            except Exception as e:
                print(f"⚠️ Failed using template {template_id}: {e}")
                continue

        if not found:
            print(f"⚠️ Manual validation required for: {field}")
            os.system("start cmd /k uvicorn Manual_Validation.main:app --reload")
            break  # Pause here to allow user to label in the UI

    return required_fields

In [ ]:
import numpy as np
from pdf2image import convert_from_path
from PIL import Image
import easyocr
import matplotlib.pyplot as plt

# ---------- SETTINGS ----------
pdf_path = "C:\\Users\\Govind\\Downloads\\Work_And\\Self_Learning\\Invoice_Extraction_demo\\Invoices\\B035_CN-016337_20240906_V1127.pdf"
poppler_path = "C:\\Users\\Govind\\Downloads\\Release-24.08.0-0\\poppler-24.08.0\\Library\\bin"

# (x, y, width, height) of region to OCR (in pixels)
#crop_region = (550, 58, 1097, 370)  # Modify as needed

# ---------- STEP 1: Convert PDF to Image ----------
pil_images = convert_from_path(pdf_path, poppler_path=poppler_path, first_page=1, last_page=1)
full_image = pil_images[0].convert("RGB")
image = Image.open("C:\\Users\\Govind\\Downloads\\Work_And\\Self_Learning\\Invoice_Extraction_demo\\Invoices\\B035_CN-016337_20240906_V1127.png")

reader = easyocr.Reader(['en'], gpu=False)

x, y = 960, 748
w, h = 727, 157
cropped = image.crop((x, y, x + w, y + h))
cropped_np = np.array(cropped)  # Convert to NumPy array
results = reader.readtext(cropped_np)
#extracted[field_name] = " ".join([r[1] for r in results]).strip()


# # ---------- STEP 2: Crop Region ----------
# x, y, w, h = crop_region
# cropped_image = full_image.crop((x, y, x + w, y + h))

# # ---------- STEP 3: OCR using EasyOCR ----------

# results = reader.readtext(np.array(cropped_image))

# ---------- STEP 4: Print Extracted Text ----------
print("Extracted Text from Cropped Region:\n")

Extracted Text from Cropped Region:

TOTAL ITEMS
AUD
202.70
Freight
AUD
0.00
GST
10.00 %
202.70
AUD
20.27
Final Amount
AUD
2222.97


In [35]:
print("Extracted Text from Cropped Region:\n")
full_text = ""
for detection in results:
    text = detection[1] if isinstance(detection, (list, tuple)) else detection
    full_text += text + "\n"

print(full_text)

Extracted Text from Cropped Region:

TOTAL ITEMS
AUD
202.70
Freight
AUD
0.00
GST
10.00 %
202.70
AUD
20.27
Final Amount
AUD
2222.97



In [ ]:
FIELD_REGEX = {
    "invoice_number": r"(CN-\d+)",
    "date": r"(\d{2}/\d{2}/\d{4})",

    # Match a pattern like: TOTAL ITEMS \n AUD \n 202.70
    "Total_Items": r"TOTAL\s*ITEMS(?:\s|\n|\r)+(?:[A-Z]+\s*)*([-+]?\d+\.\d{2})",

    # Match: Freight \n AUD \n 0.00
    "Freight": r"Freight(?:\s|\n|\r)+(?:[A-Z]+\s*)*([-+]?\d+\.\d{2})",

    # Match: Final Amount \n AUD \n 2222.97
    "Final_Amount": r"Final\s*Amount(?:\s|\n|\r)+(?:[A-Z]+\s*)*([-+]?\d+\.\d{2})",

    # Match: GST \n 10.00 % \n ... \n AUD \n 20.27
    # Extract the actual GST amount (not percent) if possible
    "GST": r"GST\s*(\d+\.\d{2})\s*%\s*[\d\s.]*AUD\s*([-+]?\d+\.\d{2})"
}


In [68]:
text = full_text.replace("\n"," ")

In [69]:
text

'TOTAL ITEMS AUD 202.70 Freight AUD 0.00 GST 10.00 % 202.70 AUD 20.27 Final Amount AUD 2222.97 '

In [72]:
re.search(FIELD_REGEX["GST"], text, re.DOTALL).group(1)

'10.00'

In [57]:
full_text

'TOTAL ITEMS\nAUD\n202.70\nFreight\nAUD\n0.00\nGST\n10.00 %\n202.70\nAUD\n20.27\nFinal Amount\nAUD\n2222.97\n'

In [ ]:
text = "TOTAL ITEMS AUD 202.70 Freight AUD 0.00"
match = re.search(r"TOTAL\s*ITEMS(?:\s|\n|\r)+(?:[A-Z]+\s*)*([-+]?\d+\.\d{2})", text, re.DOTALL)
if match:
    print("✅ Found:", match.group(1))
else:
    print("❌ Not found")


❌ Not found


In [ ]:
re.search(r"TOTAL\s*ITEMS(?:\s|\n|\r)+(?:[A-Z]+\s*)*([-+]?\d+\.\d{2})", text, re.IGNORECASE)

<re.Match object; span=(0, 22), match='TOTAL ITEMS AUD 202.70'>

### Coordinate findings and PDF processing

In [22]:
from doctr.models import ocr_predictor
from doctr.io import DocumentFile

# Load image
doc = DocumentFile.from_images("CN-006991.png")

# Load pretrained OCR model
model = ocr_predictor(pretrained=True)

# Perform OCR
result = model(doc).export()

# Access lines & their boxes
for block in result['pages'][0]['blocks']:
    for line in block['lines']:
        line_text = ' '.join([word['value'] for word in line['words']])
        print(f"Line text: {line_text}")
        print(f"Bounding box: {line['geometry']}")


Line text: INTEGRIA HEALTHCARE (AUSTRALIA) PTY LIMITED
Bounding box: ((0.333984375, 0.04850732185990336), (0.5419921875, 0.0650758605072464))
Line text: ABN 70096496212
Bounding box: ((0.3349609375, 0.06921799516908211), (0.4072265625, 0.08440582226247989))
Line text: Building 5 Freeway Office Park
Bounding box: ((0.333984375, 0.08854795692431561), (0.451171875, 0.10649720712560384))
Line text: 2728 Logan Road
Bounding box: ((0.333984375, 0.10235507246376813), (0.404296875, 0.12030432266505636))
Line text: Eight Mile Plains QLD 4113
Bounding box: ((0.333984375, 0.11478147644927539), (0.4375, 0.13273072665056362))
Line text: AUS
Bounding box: ((0.3330078125, 0.12858859198872785), (0.3544921875, 0.14515713063607083))
Line text: Ph. 1300 654 336 Fax 1300 654 844
Bounding box: ((0.3330078125, 0.1410149959742351), (0.4716796875, 0.1575835346215781))
Line text: Customer Care Ph. : 3 C
Bounding box: ((0.587890625, 0.04850732185990336), (0.6943359375, 0.06369514895330114))
Line text: Customer 

In [13]:
import fitz  # PyMuPDF
from doctr.io import DocumentFile
from doctr.models import ocr_predictor
from reportlab.pdfgen import canvas
from reportlab.lib.pagesizes import letter
from io import BytesIO

# Load model
model = ocr_predictor(pretrained=True)

# Load PDF
pdf_path = 'C:\\Users\\Govind\\Downloads\\Work_And\\Self_Learning\\Invoice_Extraction_demo\\Invoices\\CN-006991.pdf'
doc = fitz.open(pdf_path)

output_pdf = fitz.open()

# OCR + overlay text
for page_num in range(len(doc)):
    # Render page as image
    pix = doc[page_num].get_pixmap(dpi=300)
    img_bytes = pix.tobytes("png")

    # Do OCR
    image = DocumentFile.from_images(img_bytes)
    result = model(image)

    # Get dimensions
    page = doc[page_num]
    width, height = page.rect.width, page.rect.height

    # Make invisible text layer
    packet = BytesIO()
    can = canvas.Canvas(packet, pagesize=(width, height))
    
    # Draw text at OCR positions
    for block in result.export()['pages'][0]['blocks']:
        for line in block['lines']:
            for word in line['words']:
                text = word['value']
                (x_min, y_min), (x_max, y_max) = word['geometry']
                x = x_min * width
                y = height - y_max * height  # Reportlab origin is bottom-left
                font_size = (y_max - y_min) * height * 0.8
                can.setFillColorRGB(1,1,1,0)  # Invisible text
                can.setFont("Helvetica", font_size)
                can.drawString(x, y, text)

    can.save()
    packet.seek(0)

    overlay_pdf = fitz.open("pdf", packet.read())

    # Create new page
    new_page = output_pdf.new_page(width=width, height=height)
    new_page.show_pdf_page(new_page.rect, doc, pno=page_num)
    new_page.show_pdf_page(new_page.rect, overlay_pdf, 0)

# Save output
output_pdf.save('output_text_embedded.pdf')


In [ ]:
import fitz  # PyMuPDF

doc = fitz.open("output_text_embedded.pdf")

for page_num, page in enumerate(doc, start=1):
    print(f"\n--- Page {page_num} ---")
    lines = page.get_text("text").split('\n')
    for line in lines:
        print(line)


--- Page 1 ---
INTEGRIA HEALTHCARE(AUSTRALIA)PTY LIMITED
ABN 70096496212
Building5 FreewayOffice Park
2728LoganRoad
EightMilePlainsQLD 4113
AUS
Ph. 1300 654 336 Fax
1300 654 844
Customer Care Ph. :
Customer Care
Fax :
1300
Email Orders :
orders@integra.com
Integria
HEALTHCARE
Credit adjustment
note
Document
No:
Date:
Account
No:
Customer
Ref:
Delivery No:
Our Order No:
Taken By:
CN-006991
05/10/2022
55327
Ship To:
CWH Glen Waverley B086
263 Springvale Road
GLEN WAVERLEY VIC 3150
AUS
Payer:
CW Management Pty Ltd B655
BuildingF, 44 RaglanStreet
PRESTON VIC 3072
AUS
SO-0137156
Email
Material
TMIRONO
TMB12TA
TMTJS30
TMTJS30
TMOLIV60
TMDIGES
Brand
Thompsons
Thompsons
Thompsons
Thompsons
Thompsons
Thompsons
Description
OrganicIron24mg30 Tablets
Ultra B12 1000mcg100 Tablets
Turmeric Joint Support30 Tablets
Turmeric Joint Support30 Tablets
One-a-day Olive Leaf 5000mg60 Capsules
Digestion Support60 Capsules
UOM
Ordered
Shipped
EA
-1
-1
EA
1
1
EA
-2
-2
EA
-2
-2
EA
-2
-2
EA
-3
-3
RRP
List
Discou

In [10]:
import pdfplumber

with pdfplumber.open(r"C:\Users\Govind\Downloads\Work_And\Self_Learning\Invoice_Extraction_demo\Fashionova\B158_91060_20230815_V1916.pdf") as pdf:
    page = pdf.pages[0]
    words = page.extract_words()

    # Example: print all words with their bounding boxes
    for word in words:
        print(f"Text: {word['text']} | x0: {word['x0']} | top: {word['top']} | x1: {word['x1']} | bottom: {word['bottom']}")

    # Group words into lines based on 'top'
    lines = {}
    for word in words:
        key = round(word['top'])  # group by rounded top position
        lines.setdefault(key, []).append(word)

    # Now compute line bounding boxes
    for top, line_words in lines.items():
        x0 = min(w['x0'] for w in line_words)
        x1 = max(w['x1'] for w in line_words)
        bottom = max(w['bottom'] for w in line_words)
        text = " ".join(w['text'] for w in sorted(line_words, key=lambda w: w['x0']))
        print(f"Line: {text} | x0: {x0} | top: {top} | x1: {x1} | bottom: {bottom}")
        print("Words extracted:", len(words))


Text: Bill | x0: 27.838344428142 | top: 114.56623948227195 | x1: 44.24668668754199 | bottom: 125.101478937215
Text: To | x0: 49.96905653137898 | top: 114.56623948227195 | x1: 62.825946419926176 | bottom: 125.101478937215
Text: : | x0: 71.57284691787223 | top: 114.56623948227195 | x1: 75.08214555138552 | bottom: 125.101478937215
Text: CW | x0: 83.79743074632694 | top: 114.56623948227195 | x1: 101.4282313885724 | bottom: 125.101478937215
Text: Management | x0: 104.24199335598396 | top: 114.56623948227195 | x1: 169.08497981846827 | bottom: 125.101478937215
Text: Pty | x0: 172.00412612922858 | top: 114.56623948227195 | x1: 188.40192995429368 | bottom: 125.101478937215
Text: Ltd | x0: 191.07869227535184 | top: 114.56623948227195 | x1: 207.41326549440768 | bottom: 125.101478937215
Text: B655 | x0: 210.36402710817265 | top: 114.56623948227195 | x1: 235.61411577453254 | bottom: 125.101478937215
Text: Customer | x0: 370.79416362485324 | top: 117.44697223306991 | x1: 420.1034978777322 | bottom: 

In [44]:
import pdfplumber

def group_words_into_lines(words, y_tolerance=2.0):
    """
    Group words into lines by their top position, within a y_tolerance.
    """
    lines = []
    
    # Sort words top to bottom, then left to right
    words_sorted = sorted(words, key=lambda w: (w['top'], w['x0']))

    current_line = []
    current_top = None

    for word in words_sorted:
        word_top = word['top']
        
        if current_line == []:
            # Start a new line
            current_line = [word]
            current_top = word_top
        else:
            if abs(word_top - current_top) <= y_tolerance:
                current_line.append(word)
            else:
                # Finish the current line
                lines.append(current_line)
                # Start a new line
                current_line = [word]
                current_top = word_top
    
    # Add last line
    if current_line:
        lines.append(current_line)

    return lines

# Open the PDF
with pdfplumber.open(r"C:\Users\Govind\Downloads\Work_And\Self_Learning\Invoice_Extraction_demo\Fashionova\B158_91060_20230815_V1916.pdf") as pdf:
    page = pdf.pages[0]
    words = page.extract_words()

    # Group words into lines
    lines = group_words_into_lines(words)

    # Output line info
    for line_words in lines:
        x0 = min(w['x0'] for w in line_words)
        x1 = max(w['x1'] for w in line_words)
        top = min(w['top'] for w in line_words)
        bottom = max(w['bottom'] for w in line_words)
        text = " ".join(w['text'] for w in line_words)
        print(f"Line: '{text}' | x0: {x0:.2f} | top: {top:.2f} | x1: {x1:.2f} | bottom: {bottom:.2f}")

Line: 'Bill To : CW Management Pty Ltd B655' | x0: 27.84 | top: 114.57 | x1: 235.61 | bottom: 125.10
Line: 'Customer No. : PH1871' | x0: 370.79 | top: 117.45 | x1: 508.61 | bottom: 127.98
Line: 'Invoice No. : 91060' | x0: 370.79 | top: 133.65 | x1: 499.76 | bottom: 144.18
Line: 'Ship To : Chemist Warehouse Townsville B158' | x0: 27.84 | top: 137.49 | x1: 271.91 | bottom: 148.02
Line: 'Invoice Date : 15.08.23' | x0: 370.79 | top: 149.97 | x1: 511.45 | bottom: 160.50
Line: '126-128 Sturt St Address :' | x0: 28.80 | top: 157.65 | x1: 165.02 | bottom: 169.15
Line: 'Due Date : 30.09.23' | x0: 370.79 | top: 166.29 | x1: 511.45 | bottom: 176.82
Line: 'TOWNSVILLE QLD 4810' | x0: 85.32 | top: 171.09 | x1: 206.26 | bottom: 181.63
Line: 'Invoice Amount : $ 187.25 AUD AUSTRALIA' | x0: 85.32 | top: 182.49 | x1: 558.13 | bottom: 194.94
Line: 'TTAAXX IINNVVOOIICCEE' | x0: 265.92 | top: 230.83 | x1: 341.31 | bottom: 245.19
Line: 'PPaaggee 11 ooff 11' | x0: 494.16 | top: 237.61 | x1: 566.24 | bottom: 2

## Pdf_to_text_embedded

In [ ]:
import fitz  # PyMuPDF
from pdf2image import convert_from_path
import pytesseract
from PIL import Image
import io
import cv2
import numpy as np

def convert_image_pdf_to_text_pdf(input_pdf_path, output_pdf_path):
    # Set Tesseract & Poppler paths
    poppler_path = r'C:\Users\Govind\Downloads\poppler-24.08.0\Library\bin'
    #pytesseract.pytesseract.tesseract_cmd = r'C:\Program Files\Tesseract-OCR\tesseract.exe'

    # Convert PDF to images
    images = convert_from_path(input_pdf_path, poppler_path=poppler_path)
    output_pdf = fitz.open()

    for page_num, image in enumerate(images):
        # Preprocess image using OpenCV
        cv_image = cv2.cvtColor(np.array(image), cv2.COLOR_RGB2BGR)
        gray = cv2.cvtColor(cv_image, cv2.COLOR_BGR2GRAY)
        _, thresh = cv2.threshold(gray, 180, 255, cv2.THRESH_BINARY + cv2.THRESH_OTSU)
        processed_image = Image.fromarray(thresh)

        # Perform OCR
        ocr_data = pytesseract.image_to_data(processed_image, output_type=pytesseract.Output.DICT)

        # Convert original (non-thresholded) image to PNG byte stream
        img_byte_arr = io.BytesIO()
        image.save(img_byte_arr, format='PNG')
        img_byte_arr.seek(0)

        width, height = image.size
        page = output_pdf.new_page(width=width, height=height)
        page.insert_image(page.rect, stream=img_byte_arr.read())

        # Insert OCR text at real positions
        n_boxes = len(ocr_data['level'])
        for i in range(n_boxes):
            text = ocr_data['text'][i]
            conf = int(ocr_data['conf'][i])
            x, y, w, h = (
                ocr_data['left'][i],
                ocr_data['top'][i],
                ocr_data['width'][i],
                ocr_data['height'][i]
            )

            if text.strip() != '' and conf > 30:
                # Adjust y-position slightly for vertical alignment
                page.insert_text(
                    fitz.Point(x, y + h),
                    text,
                    fontsize=10,
                    fontname="helv",
                    fill=(0, 0, 0),
                    render_mode=3  # invisible text
                )
            else:
                # For debug: print what was skipped
                print(f"[SKIPPED] Conf: {conf} Text: '{text}' at ({x}, {y})")

    # Save output
    output_pdf.save(output_pdf_path)
    output_pdf.close()
    print(f"\n✅ Output PDF saved to: {output_pdf_path}")

# 🧪 Example usage:
convert_image_pdf_to_text_pdf(
    r"C:\Users\Govind\Downloads\Work_And\Self_Learning\Invoice_Extraction_demo\Invoices\CN-006991.pdf",
    r"C:\Users\Govind\Downloads\Work_And\Self_Learning\Invoice_Extraction_demo\Invoices\CN-006991_1.pdf"
)

[SKIPPED] Conf: -1 Text: '' at (0, 0)
[SKIPPED] Conf: -1 Text: '' at (96, 548)
[SKIPPED] Conf: -1 Text: '' at (96, 548)
[SKIPPED] Conf: -1 Text: '' at (96, 548)
[SKIPPED] Conf: 95 Text: ' ' at (96, 548)
[SKIPPED] Conf: -1 Text: '' at (1276, 987)
[SKIPPED] Conf: -1 Text: '' at (1276, 987)
[SKIPPED] Conf: -1 Text: '' at (1276, 987)
[SKIPPED] Conf: 95 Text: ' ' at (1276, 987)
[SKIPPED] Conf: -1 Text: '' at (92, 86)
[SKIPPED] Conf: -1 Text: '' at (92, 86)
[SKIPPED] Conf: -1 Text: '' at (788, 86)
[SKIPPED] Conf: -1 Text: '' at (486, 100)
[SKIPPED] Conf: -1 Text: '' at (787, 154)
[SKIPPED] Conf: -1 Text: '' at (786, 176)
[SKIPPED] Conf: -1 Text: '' at (787, 197)
[SKIPPED] Conf: -1 Text: '' at (786, 219)
[SKIPPED] Conf: -1 Text: '' at (787, 241)
[SKIPPED] Conf: -1 Text: '' at (92, 246)
[SKIPPED] Conf: -1 Text: '' at (1731, 307)
[SKIPPED] Conf: -1 Text: '' at (142, 319)
[SKIPPED] Conf: -1 Text: '' at (292, 352)
[SKIPPED] Conf: -1 Text: '' at (292, 374)
[SKIPPED] Conf: -1 Text: '' at (291, 405)